internal/topology/check/compatibility.go

```diff
@@ -18,39 +18,173 @@ limitations under the License.
 package check
 
 import (
-	"github.com/pkg/errors"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// ReferencedObjectsAreStrictlyCompatible checks if two referenced objects are strictly compatible, meaning that
+// ObjectsAreStrictlyCompatible checks if two referenced objects are strictly compatible, meaning that
 // they are compatible and the name of the objects do not change.
-func ReferencedObjectsAreStrictlyCompatible(current, desired client.Object) error {
+func ObjectsAreStrictlyCompatible(current, desired client.Object) field.ErrorList {
+	var allErrs field.ErrorList
 	if current.GetName() != desired.GetName() {
-		return errors.Errorf("invalid operation: it is not possible to change the name of %s/%s from %s to %s",
-			current.GetObjectKind().GroupVersionKind(), current.GetName(), current.GetName(), desired.GetName())
+		allErrs = append(allErrs, field.Forbidden(
+			field.NewPath("metadata", "name"),
+			fmt.Sprintf("cannot be changed from %v to %v to prevent incompatible changes in %v/%v", current.GetName(), desired.GetName(), current.GetObjectKind().GroupVersionKind().GroupKind().String(), current.GetName()),
+		))
 	}
-	return ReferencedObjectsAreCompatible(current, desired)
+	allErrs = append(allErrs, ObjectsAreCompatible(current, desired)...)
+	return allErrs
 }
 
-// ReferencedObjectsAreCompatible checks if two referenced objects are compatible, meaning that
+// ObjectsAreCompatible checks if two referenced objects are compatible, meaning that
 // they are of the same GroupKind and in the same namespace.
-func ReferencedObjectsAreCompatible(current, desired client.Object) error {
+func ObjectsAreCompatible(current, desired client.Object) field.ErrorList {
+	var allErrs field.ErrorList
+
 	currentGK := current.GetObjectKind().GroupVersionKind().GroupKind()
 	desiredGK := desired.GetObjectKind().GroupVersionKind().GroupKind()
-
-	if currentGK.String() != desiredGK.String() {
-		return errors.Errorf("invalid operation: it is not possible to change the GroupKind of %s/%s from %s to %s",
-			current.GetObjectKind().GroupVersionKind(), current.GetName(), currentGK, desiredGK)
+	if currentGK.Group != desiredGK.Group {
+		allErrs = append(allErrs, field.Forbidden(
+			field.NewPath("metadata", "apiVersion"),
+			fmt.Sprintf("group cannot be changed from %v to %v to prevent incompatible changes in %v/%v", currentGK.Group, desiredGK.Group, currentGK.String(), current.GetName()),
+		),
```

Comment:
nit can we move those two brackes in the same line? (same in l.62 and further down the file))